ert3/data/_record.py

```diff
@@ -0,0 +1,70 @@
+from pydantic import BaseModel, root_validator
+from typing import Mapping, Union, List, Tuple
+
+
+class _DataElement(BaseModel):
+    validate_all = True
+    validate_assignment = True
+    extra = "forbid"
+    allow_mutation = False
+    arbitrary_types_allowed = True
+
+
+class Record(_DataElement):
+    data: Union[List[float], Mapping[int, float], Mapping[str, float]]
+    index: Union[Tuple[int, ...], Tuple[str, ...]]
+
+    def __init__(self, *, data, index=None, **kwargs):
+        if index is not None:
+            pass
```

Comment:
Well, then I need to move the condition that index is None into both of the other statements... Can do that if that reads better!

Reply:
But the statement as is does nothing? since line 20 is an `if` and not `elif` the check on line 18 and 19 will be evaluated and not impact the rest of the code.

Reply:
Side note, maybe switch accounts? 😂

Reply:
You are absolutely right! It was the start of realising that since Pydantic objects have to be possible to create with all the fields explicitly, ie passing the index also due to serialisation and such, the index passed in might be inconsistent with what it ought to be... Not only did I fail to realise the realisation into code, the tiny trace of it was buggy 😂 

Anyway, I've written tests for all of the "deduced" fields now to verify consistency with the rest of the data. It resides in the "fixup" commit, but I'll squash it into the commit that introduces records before merging.

Reply:
Changes looks good! 😄